File: Libro/src/main/java/com/mycompany/libro/Nodo.java
package com.mycompany.libro;

import java.io.*;
import java.net.*;
import java.util.*;

public class Nodo {

    public static void main(String[] args) {
        try {
            // Conectar al Maestro en el puerto 5000
            Socket socket = new Socket("localhost", 3000);
            BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            PrintWriter out = new PrintWriter(socket.getOutputStream(), true);

            // Recibir fragmento del Maestro (con múltiples líneas)
            StringBuilder fragmentoCompleto = new StringBuilder();
            String linea;
            while (!(linea = in.readLine()).equals("END")) {
                fragmentoCompleto.append(linea).append("\n");
            }
            String fragmento = fragmentoCompleto.toString();
            System.out.println("Fragmento recibido: " + fragmento);

            // Procesar el fragmento con IA
            String resultado = procesarTexto(fragmento);

            // Enviar el resultado (matrices de pesos) al Maestro
            out.println(resultado);
            System.out.println("Matrices de pesos enviadas al Maestro.");

            // Cerrar las conexiones
            in.close();
            out.close();
            socket.close();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static String procesarTexto(String texto) {
        // Crear instancias de Preprocessor y NeuralNetwork
        Preprocessor preprocessor = new Preprocessor();
        int[] secuencia = preprocessor.procesarTexto(texto);

        // Parámetros de la red neuronal
        int vocabSize = preprocessor.getVocabSize();
        NeuralNetwork nn = new NeuralNetwork(vocabSize, 2048, vocabSize, 0.001);

        // Entrenamiento
        int nIteraciones = 10000;  // Reducir para pruebas
        int longitudSecuencia = 5;

        for (int epoch = 0; epoch < nIteraciones; epoch++) {
            int idx = (int) (Math.random() * (secuencia.length - longitudSecuencia));
            int[] xSeq = Arrays.copyOfRange(secuencia, idx, idx + longitudSecuencia);
            int yIdx = secuencia[idx + longitudSecuencia];

            double[] x = new double[vocabSize];
            for (int i : xSeq) {
                x[i] = 1;  // One-hot encoding
            }

            double[] y = new double[vocabSize];
            y[yIdx] = 1;

            double[] hiddenLayer = nn.forwardPropagationHidden(x);
            double[] outputLayer = nn.forwardPropagationOutput(hiddenLayer);

            nn.backPropagation(x, y, hiddenLayer, outputLayer);
        }

        // Obtener las matrices de pesos
        double[][] W1 = nn.getW1();
        double[][] W2 = nn.getW2();

        // Convertir las matrices a string para enviar
        return convertirMatrizAString(W1) + "END_MATRIX\n" + convertirMatrizAString(W2) + "END_MATRIX";
    }

    private static String convertirMatrizAString(double[][] matriz) {
        StringBuilder sb = new StringBuilder();
        for (double[] fila : matriz) {
            for (double valor : fila) {
                sb.append(valor).append(" ");
            }
            sb.append("\n");
        }
        return sb.toString();
    }

    static class Preprocessor {
        private Map<String, Integer> wordToIdx = new HashMap<>();
        private int vocabSize;

        public int[] procesarTexto(String texto) {
            String[] palabras = texto.toLowerCase().split("\\s+");
            Set<String> vocabulario = new HashSet<>(Arrays.asList(palabras));

            int idx = 0;
            for (String palabra : vocabulario) {
                wordToIdx.put(palabra, idx++);
            }
            vocabSize = vocabulario.size();

            int[] secuencia = new int[palabras.length];
            for (int i = 0; i < palabras.length; i++) {
                secuencia[i] = wordToIdx.get(palabras[i]);
            }
            return secuencia;
        }

        public int getVocabSize() {
            return vocabSize;
        }
    }

    static class NeuralNetwork {
        private double[][] W1;
        private double[][] W2;
        private int inputSize, hiddenSize, outputSize;
        private double learningRate;

        public NeuralNetwork(int inputSize, int hiddenSize, int outputSize, double learningRate) {
            this.inputSize = inputSize;
            this.hiddenSize = hiddenSize;
            this.outputSize = outputSize;
            this.learningRate = learningRate;
            W1 = new double[inputSize][hiddenSize];
            W2 = new double[hiddenSize][outputSize];
            Random rand = new Random();
            for (int i = 0; i < inputSize; i++) {
                for (int j = 0; j < hiddenSize; j++) {
                    W1[i][j] = rand.nextDouble();
                }
            }
            for (int i = 0; i < hiddenSize; i++) {
                for (int j = 0; j < outputSize; j++) {
                    W2[i][j] = rand.nextDouble();
                }
            }
        }

        public double[] forwardPropagationHidden(double[] input) {
            double[] hiddenLayer = new double[hiddenSize];
            for (int j = 0; j < hiddenSize; j++) {
                for (int i = 0; i < inputSize; i++) {
                    hiddenLayer[j] += input[i] * W1[i][j];
                }
                hiddenLayer[j] = sigmoid(hiddenLayer[j]);
            }
            return hiddenLayer;
        }

        public double[] forwardPropagationOutput(double[] hiddenLayer) {
            double[] outputLayer = new double[outputSize];
            for (int k = 0; k < outputSize; k++) {
                for (int j = 0; j < hiddenSize; j++) {
                    outputLayer[k] += hiddenLayer[j] * W2[j][k];
                }
                outputLayer[k] = sigmoid(outputLayer[k]);
            }
            return outputLayer;
        }

        public void backPropagation(double[] input, double[] target, double[] hiddenLayer, double[] outputLayer) {
            double[] outputError = new double[outputSize];
            for (int k = 0; k < outputSize; k++) {
                outputError[k] = outputLayer[k] - target[k];
            }

            double[] hiddenError = new double[hiddenSize];
            for (int j = 0; j < hiddenSize; j++) {
                for (int k = 0; k < outputSize; k++) {
                    hiddenError[j] += outputError[k] * W2[j][k];
                }
                hiddenError[j] *= sigmoidDerivative(hiddenLayer[j]);
            }

            for (int j = 0; j < hiddenSize; j++) {
                for (int k = 0; k < outputSize; k++) {
                    W2[j][k] -= learningRate * outputError[k] * hiddenLayer[j];
                }
            }

            for (int i = 0; i < inputSize; i++) {
                for (int j = 0; j < hiddenSize; j++) {
                    W1[i][j] -= learningRate * hiddenError[j] * input[i];
                }
            }
        }

        private double sigmoid(double x) {
            return 1.0 / (1.0 + Math.exp(-x));
        }

        private double sigmoidDerivative(double x) {
            return x * (1 - x);
        }

        public double[][] getW1() {
            return W1;
        }

        public double[][] getW2() {
            return W2;
        }
    }
}

File: nodo.py
import socket
import random
import numpy as np

class Preprocessor:
    def __init__(self):
        self.word_to_idx = {}
        self.vocab_size = 0

    def procesar_texto(self, texto):
        palabras = texto.lower().split()
        vocabulario = set(palabras)
        for idx, palabra in enumerate(vocabulario):
            self.word_to_idx[palabra] = idx
        self.vocab_size = len(vocabulario)
        secuencia = [self.word_to_idx[palabra] for palabra in palabras]
        return secuencia

    def get_vocab_size(self):
        return self.vocab_size


class NeuralNetwork:
    def __init__(self, input_size, hidden_size, output_size, learning_rate):
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.output_size = output_size
        self.learning_rate = learning_rate
        self.W1 = np.random.rand(input_size, hidden_size)
        self.W2 = np.random.rand(hidden_size, output_size)

    def forward_propagation_hidden(self, x):
        return self.sigmoid(np.dot(x, self.W1))

    def forward_propagation_output(self, hidden_layer):
        return self.sigmoid(np.dot(hidden_layer, self.W2))

    def back_propagation(self, x, y, hidden_layer, output_layer):
        output_error = output_layer - y
        hidden_error = np.dot(output_error, self.W2.T) * self.sigmoid_derivative(hidden_layer)

        self.W2 -= self.learning_rate * np.dot(hidden_layer.T, output_error)
        self.W1 -= self.learning_rate * np.dot(x.T, hidden_error)

    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

    def sigmoid_derivative(self, x):
        return x * (1 - x)

    def get_weights(self):
        return self.W1, self.W2


def convertir_matriz_a_string(matriz):
    return '\n'.join(' '.join(map(str, fila)) for fila in matriz)


def procesar_texto(fragmento):
    preprocessor = Preprocessor()
    secuencia = preprocessor.procesar_texto(fragmento)

    vocab_size = preprocessor.get_vocab_size()
    nn = NeuralNetwork(vocab_size, 2048, vocab_size, 0.001)

    n_iteraciones = 1000
    longitud_secuencia = 5

    for epoch in range(n_iteraciones):
        idx = random.randint(0, len(secuencia) - longitud_secuencia - 1)
        x_seq = secuencia[idx: idx + longitud_secuencia]
        y_idx = secuencia[idx + longitud_secuencia]

        x = np.zeros((1, vocab_size))
        x[0, x_seq] = 1

        y = np.zeros((1, vocab_size))
        y[0, y_idx] = 1

        hidden_layer = nn.forward_propagation_hidden(x)
        output_layer = nn.forward_propagation_output(hidden_layer)

        nn.back_propagation(x, y, hidden_layer, output_layer)

    W1, W2 = nn.get_weights()
    return convertir_matriz_a_string(W1) + "\nEND_MATRIX\n" + convertir_matriz_a_string(W2) + "\nEND_MATRIX"


def nodo():
    try:
        # Conectar al Maestro en el puerto 3000
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.connect(('localhost', 3000))
            print("Conectado al maestro")

            # Recibir fragmento del Maestro
            fragmento = ""
            while True:
                data = s.recv(1024).decode('utf-8')
                if "END" in data:
                    break
                fragmento += data

            print(f"Fragmento recibido:\n{fragmento}")

            # Procesar el fragmento con IA
            resultado = procesar_texto(fragmento)

            # Enviar el resultado (matrices de pesos) al Maestro
            s.sendall(resultado.encode('utf-8'))
            print("Matrices de pesos enviadas al maestro.")

    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    nodo()

